File: src/test/c/CITest_local_consumer.c
// consumer.c: Consumer that receives events and logs them to a file
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "et.h"

int main() {
    // Info about whole ET sytem
    et_sys_id id; et_openconfig openconfig;     // For opening existing ET system
    // Info for consumer station
    et_statconfig sconfig;
    et_stat_id    stat_id_consumer; 
    et_att_id     attach_id_consumer; // Attachment id for consumer

    int status;

    // Open an existing ET system by name (local shared memory file)
    et_open_config_init(&openconfig);
    status = et_open(&id, "/tmp/et_ci_localtest", openconfig);
    et_open_config_destroy(openconfig);
    if (status != ET_OK) {
        fprintf(stderr, "ERROR: Cannot open ET system (status %d)\n", status);
        exit(1);
    }

    //Create consumer station configuration
    et_station_config_init(&sconfig);
    et_station_config_setblock(sconfig, ET_STATION_BLOCKING);
    et_station_config_setselect(sconfig, ET_STATION_SELECT_ALL);
    et_station_config_setuser(sconfig, 1);  // Allow 1 user (consumer)
    et_station_config_setrestore(sconfig, ET_STATION_RESTORE_OUT);
    
    // Create the consumer station
    status = et_station_create(id, &stat_id_consumer, "consumer_station", sconfig);
    if (status != ET_OK) {
        fprintf(stderr, "ERROR: et_station_create failed (status %d)\n", status);
        et_close(id);
        exit(1);
    }
    et_station_config_destroy(sconfig);

    // Attach this consumer to the station
    status = et_station_attach(id, stat_id_consumer, &attach_id_consumer);
    if (status != ET_OK) {
        fprintf(stderr, "ERROR: et_station_attach (consumer) failed (status %d)\n", status);
        et_close(id);
        exit(1);
    }
    else printf("Consumer: attached to station '%s' with attach_id %d\n", "consumer_station", attach_id_consumer);

    // Open output file for logging received messages
    FILE *fout = fopen("et_output.txt", "w");
    if (!fout) {
        perror("fopen et_output.txt");
        // (Continue even if file open fails, just print to console)
    }

    printf("Consumer: attached to station, waiting for events...\n");
    // Receive 500 events and process them
    for (int count = 1; count <= 500; ++count) {
        et_event *pe;
        // Block until an event is available:contentReference[oaicite:10]{index=10}
        status = et_event_get(id, attach_id_consumer, &pe, ET_SLEEP, NULL);
        if (status != ET_OK) {
            fprintf(stderr, "ERROR: et_event_get failed at event %d (status %d)\n", count, status);
            exit(1);
        }
        // Access the event data (assume it's a null-terminated string)
        char *ev_data = (char *) pe->pdata;
        printf("Received: %s\n", ev_data);
        if (fout) {
            fprintf(fout, "%s\n", ev_data);
        }
        // Return the event to ET system (mark as processed)
        status = et_event_put(id, attach_id_consumer, pe);
        if (status != ET_OK) {
            fprintf(stderr, "ERROR: et_event_put (consumer) failed at %d (status %d)\n", count, status);
            exit(1);
        }
    }

    if (fout) fclose(fout);
    et_station_detach(id, attach_id_consumer);
    et_close(id);
    printf("Consumer: completed receiving events and disconnected.\n");
    return 0;
}

File: src/test/java/CITestLocalConsumer.java
import org.jlab.coda.et.*;
import org.jlab.coda.et.enums.Mode;
import org.jlab.coda.et.enums.Modify;
import java.io.*;
import java.nio.ByteBuffer;

public class CITestLocalConsumer {
    public static void main(String[] args) {

        // Parse arg(s)
        boolean noSocket = false;
        for (int i = 0; i < args.length; i++) {
            if (args[i].equalsIgnoreCase("-ns")) noSocket = true;
        }
        
        try {
            // Configure to open the same ET system (by file name)
            EtSystemOpenConfig config = new EtSystemOpenConfig();
            config.setEtName("/tmp/et_ci_localtest");
            config.setHost(EtConstants.hostLocal);
            config.setNetworkContactMethod(EtConstants.direct);

            
            EtSystem etSys = new EtSystem(config);
            etSys.open();  // connect to existing ET system
            // Check if using JNI library
            if (etSys.usingJniLibrary()) {
                System.out.println("ET is local\n");
            }
            else {
                if(noSocket) {
                    System.out.println("No JNI library loaded and noSocket specified. Quitting!");
                    return;
                }
                System.out.println("No JNI lib: ET running in remote mode\n");
                config.setConnectRemotely(true);
                config.setTcpPort(EtConstants.serverPort);
            }

            // Create a new station configuration for the consumer station
            EtStationConfig sConfig = new EtStationConfig();
            sConfig.setBlockMode(EtConstants.stationBlocking);
            sConfig.setSelectMode(EtConstants.stationSelectAll);
            sConfig.setUserMode(EtConstants.stationUserSingle);    // only 1 attachment (the consumer)
            sConfig.setRestoreMode(EtConstants.stationRestoreOut);

            // Create the "consumer_station" in the ET system
            EtStation station = etSys.createStation(sConfig, "consumer_station");
            // Attach this consumer process to the new station
            EtAttachment att = etSys.attach(station);
            System.out.printf("Consumer: attached to station '%s' with attach_id %d\n",
                               "consumer_station", att.getId());

            // Open output file for logging received messages
            PrintWriter outFile = null;
            try {
                outFile = new PrintWriter(new FileWriter("et_output.txt"));
            } catch (IOException io) {
                System.err.println("Warning: couldn't open et_output.txt for writing.");
            }

            System.out.println("Consumer: attached to station, waiting for events...");
            // Receive 500 events and process them
            for (int count = 1; count <= 500; ++count) {
                // Wait for an event from the station (blocking)
                EtEvent[] events = etSys.getEvents(att, Mode.SLEEP, Modify.NOTHING, 0, 1);
                if (events.length != 1) {
                    System.err.printf("ERROR: et_getEvents failed at event %d\n", count);
                    break;
                }
                EtEvent ev = events[0];
                ByteBuffer buf = ev.getDataBuffer();    // works in both socket & JNI paths
                ByteBuffer readbuf = buf.duplicate();   // duplicate to avoid changing position of original
                readbuf.order(ev.getByteOrder());       // needed since duplicate always inits in BIG_ENDIAN
                int length = ev.getLength();
                byte[] data = new byte[length];
                readbuf.get(data);

                // Read event data as a string
                if (length > 0 && data[length-1] == 0) {
                    length -= 1; // exclude null terminator for printing
                }
                String message = new String(data, 0, length, "UTF-8");
                System.out.println("Received: " + message);
                if (outFile != null) {
                    outFile.println(message);
                }
                // Mark the event as processed (put back to ET system for reuse)
                etSys.putEvents(att, events);
            }

            // Close file, detach and close ET connection
            if (outFile != null) outFile.close();
            etSys.detach(att);
            etSys.close();
            System.out.println("Consumer: completed receiving events and disconnected.");
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}
